File: src/main/java/com/dongtronic/diabot/data/A1cDTO.java
package com.dongtronic.diabot.data;

import com.dongtronic.diabot.converters.GlucoseUnit;
import org.jetbrains.annotations.Nullable;

@SuppressWarnings("Duplicates")
public class A1cDTO {
  private ConversionDTO original;
  private double dcct_mgdl;
  private double ifcc_mgdl;
  private double dcct_mmol;
  private double ifcc_mmol;


  public A1cDTO(ConversionDTO original, double dcct_mgdl, double ifcc_mgdl, @Nullable double dcct_mmol, @Nullable double ifcc_mmol) {
    this.original = original;

    this.dcct_mgdl = round(dcct_mgdl,1);
    this.ifcc_mgdl = round(ifcc_mgdl,1);
    this.dcct_mmol = round(dcct_mmol, 1);
    this.ifcc_mmol = round(ifcc_mmol, 1);
  }

  //region properties
  public ConversionDTO getOriginal() {
    return original;
  }

  public double getDcct() {
    if(original.getInputUnit() == GlucoseUnit.MMOL) {
      return dcct_mmol;
    } else if(original.getInputUnit() == GlucoseUnit.MGDL) {
      return dcct_mgdl;
    } else {
      throw new IllegalStateException();
    }
  }

  public double getIfcc() {
    if(original.getInputUnit() == GlucoseUnit.MMOL) {
      return ifcc_mmol;
    } else if(original.getInputUnit() == GlucoseUnit.MGDL) {
      return ifcc_mgdl;
    } else {
      throw new IllegalStateException();
    }
  }

  public double getDcct(GlucoseUnit unit) {
    if(unit == GlucoseUnit.MMOL) {
      return dcct_mmol;
    } else if (unit == GlucoseUnit.MGDL) {
      return dcct_mgdl;
    } else {
      throw new IllegalArgumentException();
    }
  }

  public double getIfcc(GlucoseUnit unit) {
    if(unit == GlucoseUnit.MMOL) {
      return ifcc_mmol;
    } else if (unit == GlucoseUnit.MGDL) {
      return ifcc_mgdl;
    } else {
      throw new IllegalArgumentException();
    }
  }


  //endregion

  private static double round(double value, int precision) {
    int scale = (int) Math.pow(10, precision);
    return (double) Math.round(value * scale) / scale;
  }
}

File: src/main/java/com/dongtronic/diabot/data/A1cDTO.kt
package com.dongtronic.diabot.data

import com.dongtronic.diabot.logic.diabetes.GlucoseUnit

class A1cDTO(//region properties
        val original: ConversionDTO, dcct_mgdl: Double, ifcc_mgdl: Double, dcct_mmol: Double, ifcc_mmol: Double) {
    private val dcctMgdl: Double
    private val ifccMgdl: Double
    private val dcctMmol: Double
    private val ifccMmol: Double

    val dcct: Double
        get() = when {
            original.inputUnit == GlucoseUnit.MMOL -> dcctMmol
            original.inputUnit == GlucoseUnit.MGDL -> dcctMgdl
            else -> throw IllegalStateException()
        }

    val ifcc: Double
        get() = when {
            original.inputUnit == GlucoseUnit.MMOL -> ifccMmol
            original.inputUnit == GlucoseUnit.MGDL -> ifccMgdl
            else -> throw IllegalStateException()
        }


    init {

        this.dcctMgdl = round(dcct_mgdl, 1)
        this.ifccMgdl = round(ifcc_mgdl, 1)
        this.dcctMmol = round(dcct_mmol, 1)
        this.ifccMmol = round(ifcc_mmol, 1)
    }

    fun getDcct(unit: GlucoseUnit): Double {
        return when (unit) {
            GlucoseUnit.MMOL -> dcctMmol
            GlucoseUnit.MGDL -> dcctMgdl
            else -> throw IllegalArgumentException()
        }
    }

    fun getIfcc(unit: GlucoseUnit): Double {
        return when (unit) {
            GlucoseUnit.MMOL -> ifccMmol
            GlucoseUnit.MGDL -> ifccMgdl
            else -> throw IllegalArgumentException()
        }
    }


    //endregion

    private fun round(value: Double, precision: Int): Double {
        val scale = Math.pow(10.0, precision.toDouble()).toInt()
        return Math.round(value * scale).toDouble() / scale
    }
}
